subsys/net/l2/ieee802154/ieee802154.c

```diff
@@ -119,6 +119,67 @@ static inline void set_pkt_ll_addr(struct net_linkaddr *addr, bool comp,
 	addr->type = NET_LINK_IEEE802154;
 }
 
+/**
+ * Filters the destination address of the frame (used when IEEE802154_HW_FILTER
+ * is not available).
+ */
+static bool ieeee802154_check_dst_addr(struct net_if *iface,
+				      struct ieee802154_mhr *mhr)
+{
+	struct ieee802154_context *ctx = net_if_l2_data(iface);
+
+	/*
+	 * Apply filtering requirements from chapter 6.7.2 of the IEEE
+	 * 802.15.4-2015 standard:
+	 */
+
+	if (mhr->fs->fc.dst_addr_mode == IEEE802154_ADDR_MODE_NONE) {
+		/* TODO: apply d.4 and d.5 when PAN coordinator is implemented */
+		/* also, macImplicitBroadcast is not implemented */
+		return false;
+	}
+
+	/*
+	 * c. If a destination PAN ID is included in the frame, it shall match
+	 * macPanId or shall be the broadcastPAN ID
+	 */
+	if (!(mhr->dst_addr->plain.pan_id == IEEE802154_BROADCAST_PAN_ID
+			|| mhr->dst_addr->plain.pan_id == ctx->pan_id)) {
+		LOG_WRN("Frame PAN ID does not match!");
+		return false;
+		}
+
+	if (mhr->fs->fc.dst_addr_mode == IEEE802154_ADDR_MODE_SHORT) {
+		/*
+		 * d.1. A short destination address is included in the frame,
+		 * and it matches either macShortAddress orthe broadcast
+		 * address.
+		 */
+		if (!(mhr->dst_addr->plain.addr.short_addr != IEEE802154_BROADCAST_ADDRESS ||
+				mhr->dst_addr->plain.addr.short_addr
+					!= ctx->short_addr)) {
+			LOG_WRN("Frame dst address (short) does not match!");
+			return false;
+		}
+
+	} else if (mhr->fs->fc.dst_addr_mode == IEEE802154_ADDR_MODE_EXTENDED) {
+		/*
+		 * An extended destination address is included in the frame and
+		 * matches either macExtendedAddress or, if macGroupRxMode is
+		 * set to TRUE, an 64-bit extended unique identifier (EUI-64)
+		 * group address.
+		 */
+		if (memcmp(mhr->dst_addr->plain.addr.ext_addr, ctx->ext_addr,
+					IEEE802154_EXT_ADDR_LENGTH) != 0) {
+			LOG_WRN("Frame dst address (ext) does not match!");
```

Comment:
I'm not sure if this should be at the warning level - in case there are multiple devices in the network, you'll get a flood of warnings. IMO it should be debug at most. Same for the short adress and PAN Id.